parser/c2_tokenizer.c2

```diff
@@ -1490,28 +1488,36 @@ fn bool Tokenizer.at_bol(Tokenizer* t) {
 
 fn bool Tokenizer.parse_error_warn(Tokenizer* t, Token* result, Kind kind) {
     const char* start = t.cur;
-    while (*t.cur != '\0' && *t.cur != '\r' && *t.cur != '\n')
-        t.cur++;
-    usize len = cast<usize>(t.cur - start);
-    if (len > constants.MaxErrorMsgLen) {
-        t.error(result, "error msg too long (max %d bytes)", constants.MaxErrorMsgLen);
-        return true;
+
+    // parse pptokens instead of raw text
+    string_buffer.Buf* msg = string_buffer.create_static(elemsof(t.error_msg), false, t.error_msg);
+    SrcLoc last_loc = 0;
+    while (t.lex_preproc(result) != Kind.Eof) {
+        // replace blanks with a single space
+        if (last_loc && last_loc < result.loc) msg.add1(' ');
+        // copy string text or token source
+        if (result.kind == Kind.StringLiteral) {
+            msg.add2(t.pool.idx2str(result.text_idx), result.text_len);
+        } else {
+            msg.add2(t.input_start + (result.loc - t.loc_start), result.len);
+        }
+        last_loc = result.loc + result.len;
     }
-    char[constants.MaxErrorMsgLen+1] msg;
-    string.memcpy(msg, start, len);
-    msg[len] = 0;
+    msg.size();  // ensure null terminator
```

Comment:
debug leftover?

Reply:
More anticipation of future feature :)

String buffers currently null terminate the array for each individual call. I suggest only adding the null terminator when the pointer or the size is actually needed.  This would reduce the number of redundant writes and potentially improve the inlining of single character additions.